store/hbase/snapshot.go

```diff
@@ -0,0 +1,199 @@
+// Copyright 2015 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hbasekv
+
+import (
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	"github.com/pingcap/go-hbase"
+	"github.com/pingcap/go-themis"
+	"github.com/pingcap/tidb/kv"
+)
+
+var (
+	_ kv.Snapshot     = (*hbaseSnapshot)(nil)
+	_ kv.MvccSnapshot = (*hbaseSnapshot)(nil)
+	_ kv.Iterator     = (*hbaseIter)(nil)
+)
+
+// hbaseBatchSize is used for go-themis Scanner.
+const hbaseBatchSize = 1000
+
+// hbaseSnapshot implements MvccSnapshot interface.
+type hbaseSnapshot struct {
+	txn       *themis.Txn
+	storeName string
+}
+
+// newHBaseSnapshot creates a snapshot of an HBase store.
+func newHbaseSnapshot(txn *themis.Txn, storeName string) *hbaseSnapshot {
+	return &hbaseSnapshot{
+		txn:       txn,
+		storeName: storeName,
+	}
+}
+
+// Get gets the value for key k from snapshot.
+func (s *hbaseSnapshot) Get(k kv.Key) ([]byte, error) {
+	g := hbase.NewGet([]byte(k))
+	g.AddColumn(hbaseColFamilyBytes, hbaseQualifierBytes)
+	v, err := internalGet(s, g)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return v, nil
+}
+
+// BatchGet implements kv.Snapshot.BatchGet().
+func (s *hbaseSnapshot) BatchGet(keys []kv.Key) (map[string][]byte, error) {
+	gets := make([]*hbase.Get, len(keys))
+	for i, key := range keys {
+		g := hbase.NewGet(key)
+		g.AddColumn(hbaseColFamilyBytes, hbaseQualifierBytes)
+		gets[i] = g
+	}
+	rows, err := s.txn.BatchGet(s.storeName, gets)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	m := make(map[string][]byte, len(rows))
+	for _, r := range rows {
+		k := string(r.Row)
+		v := r.Columns[hbaseFmlAndQual].Value
+		m[k] = v
+	}
+	return m, nil
+}
+
+// RangeGet implements kv.Snapshot.RangeGet().
+// The range should be [start, end] as Snapshot.RangeGet() indicated.
+func (s *hbaseSnapshot) RangeGet(start, end kv.Key, limit int) (map[string][]byte, error) {
+	scanner := s.txn.GetScanner([]byte(s.storeName), start, end, limit)
+	defer scanner.Close()
+
+	m := make(map[string][]byte)
```

Comment:
I suppose if the number of keys between start and end is significantly smaller than limit, making a big map maybe a huge waste.